Upgrading to com.unity.xr.interaction.toolkit 3.0.1 - Step 1

ms-RistoRK opened the pull request:
* Additional manual testing
Build and deployed to HL2 and tested all MRTK3 scenes; core functionality not affected negatively and no broken asset was detected.

IceTrooper commented:
I am a doing research on VR frameworks and I really liked the MRTK approach in that it builds upon XRIT rather than trying to replace it. I'm impressed with what's been done in version 3.0 of XRIT and I'd love to see this PR come out as soon as possible. Do you think it's a matter of days/weeks/months? ;)

ms-RistoRK commented:
> I am a doing research on VR frameworks and I really liked the MRTK approach in that it builds upon XRIT rather than trying to replace it. I'm impressed with what's been done in version 3.0 of XRIT and I'd love to see this PR come out as soon as possible. Do you think it's a matter of days/weeks/months? ;) @ms-RistoRK

Hello @IceTrooper, thank you for reaching out.  We have it planned to release in 3 to 4 weeks, could be earlier though (I'm trying to make it happen asap).  We are currently gathering internal feedback and doing extensive testing but everything is moving in the right direction yet we want to make sure before incorporating it into main branch.  I'll ping you once it is in main branch.

AMollis commented:
@ms-RistoRK , this doesn't seem to remove usages of XRI's deprecated Controller-based classes.  I think this will require make a new MRTK3 Rig prefab, and remove usage of all controller-based classes.

Take a look at all usages of `ActionBasedController`, `XRBaseController`, `XRController`, `XRControllerState`, and `XRScreenSpaceController`.  According to the deprecation notes, these classes have been deprecated.  In which case, we should remove MRTK3's usages of these, and deprecate prefabs and type extensions where needed.

https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@3.0/manual/whats-new-3.0.html#deprecations

ms-RistoRK commented:
> @ms-RistoRK , this doesn't seem to remove usages of XRI's deprecated Controller-based classes. I think this will require making a new MRTK3 Rig prefab and remove usage of all controller-based classes.
> 
> Take a look at all usages of `ActionBasedController`, `XRBaseController`, `XRController`, `XRControllerState`, and `XRScreenSpaceController`. According to the deprecation notes, these classes have been deprecated. In which case, we should remove MRTK3's usages of these, and deprecate prefabs and type extensions where needed.
> 
> https://docs.unity3d.com/Packages/com.unity.xr.interaction.toolkit@3.0/manual/whats-new-3.0.html#deprecations

We should deprecate MRTK3 rig entirely and build on top of Unity's rig instead.

ms-RistoRK commented:
Update: Changed PR base branch to https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/tree/feature/XRI3 as per maintainers meeting agreement.

ms-RistoRK commented:
/asp run

ms-RistoRK commented:
/azp run

shaynie commented:
/azp run

ms-RistoRK commented:
I've changed strategies on how to implement this Epic (not a User Story).  I'll be submitting individual PRs to gradually make progress instead of one gigantic PR at the end.  The reason is because removing the deprecated classes and prefabs forces me to change so many things and every time I remove one thing then 10+ new things have to be re-implemented.  This Epic (not a User Story) is definitively more than just changing Action references in prefabs.
This first PR is ready for review.
In the first PR I already touched 155 files, 60 commits and at least it removes the old Action references, upgrade packages, remove deprecated packages, upgrade prefabs, and it works in Editor and HL2.  The next PRs will focus on removing deprecated classes and re-build MRTK3 without those deprecated classes.

whebertML commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/SolverHandler.cs:
```diff
@@ -351,6 +351,7 @@ protected virtual void Update()
         /// <summary>
         /// A Unity event function that is called every frame after normal update functions, if this object is enabled.
         /// </summary>
+        [System.Obsolete]
```
nit: ` using System;` above  and `[Obsolete]` just for consistency.

ms-RistoRK replied:
I agree, I missed it in the first editing, thanks for letting me know.  Fixed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/865c7bff4727cb06d9c91f5dfe541340af63ac78 , resolving.

whebertML commented on org.mixedrealitytoolkit.spatialmanipulation/Tests/Runtime/BoundsControlTests.cs:
```diff
@@ -299,7 +299,7 @@ public IEnumerator TestNoHandlesToggleWhenMovingWithObjectManipulator([ValueSour
             Assert.IsFalse(bc.HandlesActive, "Handles should not have been toggled.");
 
-            TestUtilities.AssertAboutEqual(bc.transform.position, initialObjectPosition, $"Object should be placed generally in the same position! Actual position: {bc.transform.position:F5}, should be {initialObjectPosition}", 0.00001f);
+            TestUtilities.AssertAboutEqual(bc.transform.position, initialObjectPosition, $"Object should be placed generally in the same position! Actual position: {bc.transform.position:F5}, should be {initialObjectPosition}", 0.01f);
```
Doesn't seem related to this change, was it not passing without this?

ms-RistoRK replied:
Exactly, wasn't passing, here is a screenshot of the failure:
<img width="878" alt="image" src="https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/2c2640a4-034e-4169-a437-1419d86fc547">

whebertML commented:
> I've changed strategies on how to implement this Epic (not a User Story). I'll be submitting individual PRs to gradually make progress instead of one gigantic PR at the end. The reason is because removing the deprecated classes and prefabs forces me to change so many things and every time I remove one thing then 10+ new things have to be re-implemented. This Epic (not a User Story) is definitively more than just changing Action references in prefabs. This first PR is ready for review. In the first PR I already touched 155 files, 60 commits and at least it removes the old Action references, upgrade packages, remove deprecated packages, upgrade prefabs, and it works in Editor and HL2. The next PRs will focus on removing deprecated classes and re-build MRTK3 without those deprecated classes.

@ms-RistoRK  Totally agreed we should make changes one step at a time.  I thought the strategy regarding prefabs was to leave the prefabs as is, just mark them deprecated (either name change or move to a Deprecated folder).  This change seems to highly modify the existing prefabs, and then create new "Obsolete" prefabs that are themselves heavily modifications of the originals.  In my view, this initial change should only be primarily about compilation fixes (namespace changes, adding [Obsolete] where needed), etc.  I think we should revert any prefab modifications from this initial PR, minus perhaps the marking of Deprecated.

ms-RistoRK commented on org.mixedrealitytoolkit.core/Editor/Editors/BaseInteractableEditor.cs:
```diff
@@ -12,7 +12,8 @@ namespace MixedReality.Toolkit.Editor
     [CustomEditor(typeof(MRTKBaseInteractable), true)]
-    public class BaseInteractableEditor : XRBaseInteractableEditor
+    [CanEditMultipleObjects]
+    public class BaseInteractableEditor : UnityEditor.XR.Interaction.Toolkit.Interactables.XRBaseInteractableEditor
```
Indeed, I missed that in the first pass.  Thank you for letting me know.  Fixed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/31a11656290adbdad63f775e0b18190f65211c87 resolving this.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Experimental/SpatialMouse/Interactor/SpatialMouseInteractor.cs:
```diff
@@ -1,14 +1,13 @@
+using UnityEngine.XR.Interaction.Toolkit.Interactors;
 
 namespace MixedReality.Toolkit.Input.Experimental
 {
```
Updated in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/850b0816db28f2a1f5e9301874a1b68e0937b0ff , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Interactors/HandJointInteractor.cs:
```diff
@@ -45,6 +48,7 @@ public abstract class HandJointInteractor :
+        [Obsolete]
         public override bool isHoverActive
         {
             // Only be available for hovering if the controller is tracked or we have joint data.
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/966c9ef2c03f4f804387e1cae33509e8f9d87e9c , resolving.

MaxPalmer-UH commented:
> > I've changed strategies on how to implement this Epic (not a User Story). I'll be submitting individual PRs to gradually make progress instead of one gigantic PR at the end. The reason is because removing the deprecated classes and prefabs forces me to change so many things and every time I remove one thing then 10+ new things have to be re-implemented. This Epic (not a User Story) is definitively more than just changing Action references in prefabs. This first PR is ready for review. In the first PR I already touched 155 files, 60 commits and at least it removes the old Action references, upgrade packages, remove deprecated packages, upgrade prefabs, and it works in Editor and HL2. The next PRs will focus on removing deprecated classes and re-build MRTK3 without those deprecated classes.
> 
> @ms-RistoRK Totally agreed we should make changes one step at a time. I thought the strategy regarding prefabs was to leave the prefabs as is, just mark them deprecated (either name change or move to a Deprecated folder). This change seems to highly modify the existing prefabs, and then create new "Obsolete" prefabs that are themselves heavily modifications of the originals. In my view, this initial change should only be primarily about compilation fixes (namespace changes, adding [Obsolete] where needed), etc. I think we should revert any prefab modifications from this initial PR, minus perhaps the marking of Deprecated.
> 
> Also, I thought the strategy regarding prefabs from the last meeting was to keep the original (deprecated) prefabs intact, using deprecated components. Then, create new prefabs (not in the initial PR) that will slowly be developed to utilize the new XRI-isms that work with the rest of MRTK3's elements and components, over time.

Wyck, that matches my understanding of what was agreed - deprecate the old prefabs, in the absence of a good way of doing this in Unity. suggested by moving them to a deprecated folder.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Assets/Prefabs/MRTK XR Rig.prefab:
```diff
@@ -28,13 +28,13 @@ Transform:
   m_PrefabInstance: {fileID: 0}
   m_PrefabAsset: {fileID: 0}
   m_GameObject: {fileID: 2351505566771328562}
+  serializedVersion: 2
```
That was my original approach, however, I noticed that approach would have required me to update each scene with the new prefabs making this PR bigger so I decided for the smaller PR approach.
As for the changes of the Obsolete prefabs:
1 - I went with the Obsolete* renaming because I found a prefab that had such renaming (org.mixedrealitytoolkit.input\Assets\Prefabs\Obsolete MRTK Hand Controller.prefab) and wanted to be consistent with how it has been done before.
2 - The obsolete prefabs content was modified not by me but by the upgrade tool when upgrading to packages' versions, in particular, the com.unity.xr.interaction.toolkit 3.0.1 package modified the prefabs containing an XRController component or those in which the XRController functionality was transferred to.  For example, on the left we have the pre XRI 3 migration prefab and on the right we have the same prefab after com.unity.xr.interaction.toolkit package version was bumped to 3.0.1
<img width="491" alt="image" src="https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/c033e78c-b506-40a3-b7d3-c1d21bf5a87f">

ms-RistoRK commented:
> > > I've changed strategies on how to implement this Epic (not a User Story). I'll be submitting individual PRs to gradually make progress instead of one gigantic PR at the end. The reason is because removing the deprecated classes and prefabs forces me to change so many things and every time I remove one thing then 10+ new things have to be re-implemented. This Epic (not a User Story) is definitively more than just changing Action references in prefabs. This first PR is ready for review. In the first PR I already touched 155 files, 60 commits and at least it removes the old Action references, upgrade packages, remove deprecated packages, upgrade prefabs, and it works in Editor and HL2. The next PRs will focus on removing deprecated classes and re-build MRTK3 without those deprecated classes.
> > 
> > 
> > @ms-RistoRK Totally agreed we should make changes one step at a time. I thought the strategy regarding prefabs was to leave the prefabs as is, just mark them deprecated (either name change or move to a Deprecated folder). This change seems to highly modify the existing prefabs, and then create new "Obsolete" prefabs that are themselves heavily modifications of the originals. In my view, this initial change should only be primarily about compilation fixes (namespace changes, adding [Obsolete] where needed), etc. I think we should revert any prefab modifications from this initial PR, minus perhaps the marking of Deprecated.
> > Also, I thought the strategy regarding prefabs from the last meeting was to keep the original (deprecated) prefabs intact, using deprecated components. Then, create new prefabs (not in the initial PR) that will slowly be developed to utilize the new XRI-isms that work with the rest of MRTK3's elements and components, over time.
> 
> Wyck, that matches my understanding of what was agreed - deprecate the old prefabs, in the absence of a good way of doing this in Unity. suggested by moving them to a deprecated folder.

That was my original approach, however, I noticed that approach would have required me to update each scene with the new prefabs making this PR bigger so I decided for the smaller PR approach.
As for the changes of the Obsolete prefabs:
1 - I went with the Obsolete* renaming because I found a prefab that had such renaming (org.mixedrealitytoolkit.input\Assets\Prefabs\Obsolete MRTK Hand Controller.prefab) and wanted to be consistent with how it has been done before.
2 - The obsolete prefabs content was modified not by me but by the upgrade tool when upgrading to packages' versions, in particular, the com.unity.xr.interaction.toolkit 3.0.1 package modified the prefabs containing an XRController component or those in which the XRController functionality was transferred to. For example, on the left we have the pre XRI 3 migration prefab and on the right we have the same prefab after com.unity.xr.interaction.toolkit package version was bumped to 3.0.1
image.
<img width="491" alt="image" src="https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/cbc0285c-103e-4ee3-b6c0-01a1590a17ae">

AMollis reviewed (changes requested):
One major overall comment.  You've marked all functions and properties that utilized controllers as `[Obsolete]`.  However, we aren't 100% we cannot remove usage of the controller usage.  Instead, for this first PR, remove all the `[Obsolete]` attributes on functions and properties.  Only when we are 100% there is no alternative should we make as `[Obsolete]`.

I started marking all function and properties that you add obsolete to, however I stop at some point.  In general, remove all added `[Obsolete]` attributes you've added to functions and properties.  We'll add these later only once we're 100% there isn't a better alternative.

ms-RistoRK commented on UnityProjects/MRTKDevTemplate/ProjectSettings/boot.config:
No idea, added by Unity automatically.  I found that :/ :
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/7f303d63-f2e9-4052-900f-f002ba66b075)

I decided to add it to source control because Unity added it to ProjectSettings folder and documentation states that:
![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/968ccab3-7ac6-4d5a-9e5f-2c4ea3f63876)

whebertML replied:
My understanding is that the update in Unity version (2022.3.x and above), this file gets removed from previous installations.  We've seen this in our projects as well since ML2 requires the newer versions of Unity.

ms-RistoRK commented on org.mixedrealitytoolkit.accessibility/CHANGELOG.md:
```diff
@@ -2,7 +2,13 @@
+
+### Fixed
+
+* Reduced package description to support for UPM package publishing in the Unity Asset Store.
```
Correct, from a merge from main because in a maintainers meeting someone requested to keep the branch up to date with main but I'll do merges from main in separate PRs from now on.  Resolving.

AMollis replied:
Yes, keep the branch up-to-date...but do this with a separate PR, do not mixed with your changes.  Otherwise, it is difficult to differentiate between what you changed and what others changed.  

So next time keep this branch up-to-date with standalone merge PR, and do not add your own changes

ms-RistoRK replied:
Will do.

ms-RistoRK commented on org.mixedrealitytoolkit.input/CHANGELOG.md:
```diff
@@ -2,13 +2,21 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 
-## [3.2.1-development] - 2024-03-20
+## [4.0.0-development] - <TBD>
```
Oh shoot! I missed that, I put a TBD to avoid updating the date in each rebase from main, thx for catching that.  Fixed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/369ed49fd27ac3fa1b3b267e5c0a21bf2b81cbc7  , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+    /// <remarks>
+    /// TBD
+    /// </remarks>
+    public class MRTK3ModelXRI3 : MonoBehaviour
```
I went with 'HandModel' as I believe to be closer to its purpose.  Fixed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e8edbc7671ec4e32f0b1f79c3a1d664c5282d93f , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Tests/Runtime/SolverTapToPlaceTests.cs:
```diff
@@ -23,6 +24,8 @@ public class SolverTapToPlaceTests : BaseRuntimeInputTests
         /// </summary>
         [UnityTest]
+        [Obsolete]
+        [Ignore("Temporary ignoring while completing XRI 3 migration.  TODO: Re-implement this test once XRI 3 migration is completed.")]
```
Done https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/735 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/Solver.cs:
```diff
@@ -245,6 +246,7 @@ protected virtual void OnEnable()
         /// <summary>
         /// A Unity event function that is called when the script component has been disabled.
         /// </summary>
+        [Obsolete]
```
Done https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/737 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Tests/Runtime/TransformViaInteractorTests.cs:
```diff
@@ -26,6 +26,7 @@ public class TransformViaInteractorTests : BaseRuntimeInputTests
         /// applied to target object should be respected.
         /// </summary>
         [UnityTest]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/issues/738 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.core/CHANGELOG.md:
```diff
@@ -2,7 +2,13 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 
-## [3.2.0-development] - 2024-03-20
+## [4.0.0-development] - <TBD>
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/369ed49fd27ac3fa1b3b267e5c0a21bf2b81cbc7 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/CHANGELOG.md:
```diff
@@ -2,11 +2,21 @@
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
 
-## [3.3.0-development] - 2024-04-5
+## [4.0.0-development] - <TBD>
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/369ed49fd27ac3fa1b3b267e5c0a21bf2b81cbc7 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcomponents/CHANGELOG.md:
```diff
@@ -1,7 +1,24 @@
 
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/).
+
+## [4.0.0-development] - <TBD>
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/369ed49fd27ac3fa1b3b267e5c0a21bf2b81cbc7 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/BoundsControl/BoundsControl.cs:
```diff
@@ -406,6 +408,34 @@ public ConstraintManager ConstraintsManager
 
+        [SerializeField]
+        [Tooltip("The concrete types of ManipulationLogic<T> to use for manipulations.")]
+        private LogicType manipulationLogicTypes = new LogicType
```
Got it, will do, resolving this.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+    /// This component replaces the functionality of the deprecated XRController Model functionality to .
+    /// </summary>
+    /// <remarks>
+    /// TBD
```
Removed the TBD, kept the remarks though with a comment I consider to be pertinent.  https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/180a957e18517b8843496151acb3fdf07d9f33fd , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+    {
+        //TODO: Add Unity-test that check the MRTK3XRI3Model has the correct fields.
+
+        #region MRTK3 XRI 3 Model Properties
```
Updated in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/ca541f517706d20651736ec2c71d2ac71a423ba4 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+    /// </remarks>
+    public class MRTK3ModelXRI3 : MonoBehaviour
+    {
+        //TODO: Add Unity-test that check the MRTK3XRI3Model has the correct fields.
```
Correct, it was a reminder to add a smoke-test for MRTK3XRI3Model (now HandModel).  I had already added the smoke-test but forgot to remove the comment, sorry about that.  The smoke-test in question: https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/251d2dce9659e8928774488fd106bebbae84f654 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+        private Transform modelPrefab;
+
+        /// <summary>
+        /// The prefab of the MRTK Controller to show that will be automatically instantitated by this behaviour.
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/1b377fabd1eedfbaecbd0992438af47fb0d08c9b , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+        private Transform modelParent;
+
+        /// <summary>
+        /// The transform that is used as the parent for the model prefab when it is instantiated.  Will be set to a new child GameObject if None.
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/9301241d4ff84272f252479f4c6c7bd49c22dbfc , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+        }
+
+        // Update is called once per frame
+        void Update()
```
Done, resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+
+        // Start is called before the first frame update
+        void Start()
+        {
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/fbfa57c0377e4e1bbe0dbde22160517bc7523a76 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+        #endregion Associated hand select values
+
+        // Start is called before the first frame update
+        void Start()
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/688e08980d8378df6406f86786c1c6e4a20832f0 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/CHANGELOG.md:
```diff
@@ -2,11 +2,21 @@
+
+* Updated package com.unity.xr.interaction.toolkit to 3.0.1
+
+## [3.3.0-development] - 2024-04-30
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/CHANGELOG.md:
```diff
@@ -2,11 +2,21 @@
+
+* Added null check and index check when hiding colliders on BoundsHandleInteractable. [PR #730](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/730)
 
 ## [3.2.0-development] - 2024-03-20
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/CHANGELOG.md:
```diff
@@ -2,13 +2,21 @@
+
+* Updated package com.unity.xr.interaction.toolkit to 3.0.1
+
+## [3.2.1-development] - 2024-4-23
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.core/CHANGELOG.md:
```diff
@@ -2,7 +2,13 @@
+
+* Updated package com.unity.xr.interaction.toolkit to 3.0.1
+
+## [3.2.0-development] - 2024-04-23
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcomponents.noncanvas/CHANGELOG.md:
```diff
@@ -2,7 +2,19 @@
+
+* Updated package com.unity.xr.interaction.toolkit to 3.0.1
+
+## [3.1.3-development] - 2024-04-17
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcomponents/CHANGELOG.md:
```diff
@@ -1,7 +1,24 @@
+
+* Updated package com.unity.xr.interaction.toolkit to 3.0.1
+
+## [3.3.0-development] - 2024-04-26
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.uxcore/CHANGELOG.md:
```diff
@@ -2,7 +2,20 @@
+
+* Updated package com.unity.xr.interaction.toolkit to 3.0.1
+
+## [3.2.1-development] - 2024-04-23
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/becdbe4b3f5146a7a145483f17d711d928424362 , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Interactors/HandJointInteractor.cs:
```diff
@@ -31,6 +33,7 @@ public abstract class HandJointInteractor :
         #region IHandedInteractor
 
         /// <inheritdoc/>
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Interactors/Ray/MRTKRayInteractor.cs:
```diff
@@ -44,6 +46,7 @@ public class MRTKRayInteractor :
         /// Used to check if the parent controller is tracked or not
         /// Hopefully this becomes part of the base Unity XRI API.
         /// </summary>
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Tests/Runtime/InteractionModeManagerTests.cs:
```diff
@@ -26,6 +28,7 @@ public class InteractionModeManagerTests : BaseRuntimeInputTests
         /// Also checks that the proximity detector doesn't trigger hovers on other objects
         /// </summary>
         [UnityTest]
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Interactors/Poke/PokeInteractor.cs:
```diff
@@ -44,6 +47,7 @@ protected virtual bool TryGetPokePose(out Pose pose)
         /// Called during ProcessInteractor to obtain the poking radius. All raycasts and other physics detections
         /// are done according to this radius. Override to customize how the radius is calculated.
         /// </summary>
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Tests/Runtime/BasicInputTests.cs:
```diff
@@ -41,10 +49,43 @@ public IEnumerator InputDeviceSmoketest()
+        /// Ensures that the LeftHandController, RightHandController, and GazeController controllers have one and only one of the required components to work with com.unity.xr.interaction.toolkit 3.0.1 package
+        /// </summary>
+        [UnityTest]
+        [Obsolete] //TODO: The [Obsolete] attribute can be removed once the LeftHandController, RightHandController, and GazeController have stopped being obsolete by removing all the XRController as part of the XRI 3 migration
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Visualizers/RiggedHandVisualizer/RiggedHandMeshVisualizer.cs:
```diff
@@ -74,6 +75,7 @@ public bool ShowHandsOnTransparentDisplays
         private List<XRDisplaySubsystem> displaySubsystems = new List<XRDisplaySubsystem>();
 
         // The XRController that is used to determine the pinch strength (i.e., select value!)
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/ObjectManipulator/ObjectManipulator.cs:
```diff
@@ -1052,14 +1054,15 @@ private void ReleaseRigidBody(Vector3 velocity, Vector3 angularVelocity)
         /// Gets the absolute device (grip) rotation associated with the specified interactor.
         /// Used to query actual grabbing rotation, vs a ray rotation.
         /// </summary>
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/DirectionalIndicator.cs:
```diff
@@ -56,6 +57,7 @@ public class DirectionalIndicator : Solver
 
         /// <inheritdoc/>
+        [Obsolete]
         protected override void Start()
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/ConstantViewSize.cs:
```diff
@@ -157,6 +158,7 @@ public float FovScale
         private float objectSize = 1f;
 
         /// <inheritdoc/>
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/Follow.cs:
```diff
@@ -345,6 +346,7 @@ protected override void OnEnable()
             new ProfilerMarker("[MRTK] Follow.SolverUpdate");
 
         /// <inheritdoc />
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraint.cs:
```diff
@@ -211,6 +211,7 @@ public UnityEvent OnHandDeactivate
             new ProfilerMarker("[MRTK] HandConstraint.SolverUpdate");
 
         /// <inheritdoc />
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -144,6 +145,7 @@ public float HeadGazeProximityThreshold
         /// This method will modify the position and rotation behavior of the hand
         /// constraint if the followHandUntilFacingCamera variable is enabled.
         /// </remarks>
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/InBetween.cs:
```diff
@@ -113,6 +114,7 @@ public Transform SecondTransformOverride
         private SolverHandler secondSolverHandler;
 
         /// <inheritdoc/>
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Solvers/Momentum.cs:
```diff
@@ -96,6 +97,7 @@ protected override void OnEnable()
             new ProfilerMarker("[MRTK] Momentum.SolverUpdate");
 
         /// <inheritdoc />
+        [Obsolete]
```
Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef , resolving.

ms-RistoRK commented:
> One major overall comment. You've marked all functions and properties that utilized controllers as `[Obsolete]`. However, we aren't 100% we cannot remove usage of the controller usage. Instead, for this first PR, remove all the `[Obsolete]` attributes on functions and properties. Only when we are 100% there is no alternative should we make as `[Obsolete]`.
> 
> I started marking all function and properties that you add obsolete to, however I stop at some point. In general, remove all added `[Obsolete]` attributes you've added to functions and properties. We'll add these later only once we're 100% there isn't a better alternative.

Done in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/e65add2b344c85b36879f054c1bdeb820d33a3ef by reverting 853c987, 8cecc35, 6c128bc, and 4c3e4e2.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Tests/Runtime/ObjectManipulatorTests.cs:
```diff
@@ -212,6 +213,7 @@ public IEnumerator ObjectManipulatorInstantiate()
         /// from gaze provider.
         /// </summary>
         [UnityTest]
+        [Ignore("Temporary ignoring while completing XRI 3 migration.  TODO: Re-implement this test once XRI 3 migration is completed.")]
```
Done.

ms-RistoRK commented on org.mixedrealitytoolkit.spatialmanipulation/Tests/Runtime/SolverTapToPlaceTests.cs:
```diff
@@ -106,6 +109,8 @@ public IEnumerator TapToPlaceFollowsRightHandRay()
         /// </summary>
         [UnityTest]
+        [Obsolete]
+        [Ignore("Temporary ignoring while completing XRI 3 migration.  TODO: Re-implement this test once XRI 3 migration is completed.")]
```
Done.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Controllers/MRTK3 Model XRI3.cs:
```diff
@@ -0,0 +1,84 @@
+        // Start is called before the first frame update
+        void Start()
+        {
+            if (!HandNode.Equals(XRNode.LeftHand) && !HandNode.Equals(XRNode.RightHand))
```
I added it because of two reasons:
1 - It is part of the deprecated XRController Model functionality that needs to be replaced (Unity devs mentioned in [Slack](https://app.slack.com/client/E016WLPF0G6/C02GHQK9CP3?force_cold_boot=1) they don't plan to create a new home for it) as in:

![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/a47a01f7-7b2f-407e-a98c-4042cd024bed)
2 - Because we use it throughout the code via [ArticulatedHandController](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/blob/main/org.mixedrealitytoolkit.input/Controllers/ArticulatedHandController.cs) as in:

![image](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/assets/84108471/7d0231eb-51ca-4bad-ac41-87b76265ca22)

I thought on replicating it so that when obsolete classes/components are removed then editing its references would be easier.

ms-RistoRK commented:
In the last commits I have restored the input actions to be consumed via XRController component just as they are currently in the main branch, also reverted changes to unity-tests to make them work as they were.  I moved the prefabs with an emptied XRController component to an Experimental folder to continue working on them.  I believe to have addressed the concerns expressed in the comments, @whebertML, @AMollis, @MaxPalmer-UH if you have time, Could you please take another pass at the PR? thank you.

AMollis commented on org.mixedrealitytoolkit.input/Experimental/XRI3/HandModel.cs:
```diff
@@ -0,0 +1,78 @@
+namespace MixedReality.Toolkit.Input
+{
+    /// <summary>
+    /// This component stores a reference to the hand model to be used in Editor mode.
```
Comment is not correct.   Nothing in this class control if this is visible in Editor or not.

Change to something like
```
/// <summary>
/// This component instantiates a model prefab for hand interactors
/// </summary>
/// <remarks>
/// This does not control the visibility of the instantiated models, the prefab is always created.
/// </remarks>
```

AMollis replied:
Also, as we progress through the XR13 support, you may find this class is not need.   In future changes, let's see if you can just add the "hand model" prefab to the MRTK Rig itself.  This class is not doing anything special, just creating a game object, so I don't see the value. The only useful bit is that it stores some public properties that other components can utilize, and I wonder if this is really needed.

ms-RistoRK replied:
Updated comments in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/4f971067f6c668d55633ca0e87c398595e5b4d41, resolving.

AMollis reviewed (changes requested):
Some minor tweaks needed still.

Also, PLEASE next time pull the `main` branch commits with a separate PR.  Including the main commits here makes it more difficult to differentiate between your changes and changes from others.  

Also, a friendly reminder, don't squash your changes when you complete this PR.....If you do, future merges/rebases from `main` will have conflicts.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Experimental/XRI3.meta:
```diff
@@ -0,0 +1,8 @@
+fileFormatVersion: 2
+guid: a7771016ecaf35b45a99cb6ed122a832
+folderAsset: yes
```
Oh, that wasn't me, that is added by Unity Editor I have no control over it.  Actually, I just checked and all our .meta files have that exact line.  I suspect the field must indicate some sort of the .meta file version format but couldn't find any documentation about.
Thank you for the best-practice hint though.  Resolving.

ms-RistoRK commented on org.mixedrealitytoolkit.input/Assets/Prefabs/Experimental-XRI3.meta:
```diff
@@ -1,5 +1,6 @@
 fileFormatVersion: 2
-guid: 1b2cf9d67a560ff4faa894148c08c3e2
+guid: 621e148bde616f74e909a21ea39ff8fe
+folderAsset: yes
```
Yep, that is my line of thought too: a temporarily folder for the XRI3 migration development which will be removed once finished.  I just added the -XRI3 to easily differentiate with other Experimental folders.  I'll keep in my mind your suggestion and be sure the version does not reach main.

whebertML commented on org.mixedrealitytoolkit.core/Interactors/IGazeInteractor.cs:
```diff
@@ -15,4 +15,4 @@ public interface IGazeInteractor : IXRInteractor
     }
-}
\ No newline at end of file
+}
```
nit:  minor thing, but would it be possible to remove any of these extraneous line breaks?  There are a few throughout.

whebertML replied:
Or are these being done automatically for any files that don't already have one at the end?

ms-RistoRK replied:
I agree they are annoying but those are added by Visual Studio whenever it saves a file.  I even tried [this](https://stackoverflow.com/questions/64501728/how-to-stop-visual-studio-from-putting-extra-newline-at-the-end-of-a-file-on-sav) to make it stop but Visual Studio still adds them whenever it saves a file.  Sorry about that.

whebertML commented on org.mixedrealitytoolkit.input/Tests/Runtime/BasicInputTests.cs:
```diff
@@ -27,6 +32,9 @@ namespace MixedReality.Toolkit.Input.Tests
     /// </summary>
     public class BasicInputTests : BaseRuntimeInputTests
     {
+        private const string OPENXR_RIGHT_HAND_PREFAB_GUID = "da93d751ddc0f64468dfc02f18d02d00";
+        private const string OPENXR_LEFT_HAND_PREFAB_GUID = "2b468cc4fe6d2b44ebc53b958b38b91a";
```
What are these for, they don't appear to be used anywhere?

ms-RistoRK replied:
Oh!, remnants of Unity-tests I was using for testing, I forgot to remove them.  Thanks for catching that.  Removed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/d74062c83f752f7bf380e17a3465dce5dd6b7f99 , resolving.

whebertML commented on org.mixedrealitytoolkit.input/Tests/Runtime/BasicInputTests.cs:
```diff
@@ -41,9 +49,44 @@ public IEnumerator InputDeviceSmoketest()
+        ///// Ensures that the LeftHandController, RightHandController, and GazeController controllers have one and only one of the required components to work with com.unity.xr.interaction.toolkit 3.0.1 package
+        ///// </summary>
+        //[UnityTest]
+        //public IEnumerator XRI3ControllerHaveOneAndOnlyOneOfTheNeededXRI3ComponentsTest()
+        //{
+        //    var controllers = new[] {
+        //        CachedLookup.LeftHandController,
+        //        CachedLookup.RightHandController,
+        //        CachedLookup.GazeController
+        //    };
+
+        //    foreach (var controller in controllers)
+        //    {
+        //        Assert.That(controller, Is.Not.Null);
+
+        //        // Check all controllers have one and only one TrackedPoseDriver component
+        //        TrackedPoseDriver[] controllerTrackedPoseDrivers = controller.GetComponents<TrackedPoseDriver>();
+        //        Assert.AreEqual(controllerTrackedPoseDrivers.Length, 1);
+
+        //        // Check that LeftHandController and RightHandController also have one and only one XRRayInteractor component
+        //        if (controller.Equals(CachedLookup.LeftHandController) || controller.Equals(CachedLookup.RightHandController))
+        //        {
+        //            XRRayInteractor[] controllerXRRayInteractors = controller.GetComponents<XRRayInteractor>();
+        //            Assert.AreEqual(controllerXRRayInteractors.Length, 1);
+        //        }
+        //    }
+
+        //    yield return null;
```
Even though these new XRI3 tests are commented out, it seems perhaps premature to be writing tests for the upcoming XRI3-based rig when we don't fully know what that rig will look like.  For instance, the LeftHandController, RightHandController, GazeController, which are deprecated XRBaseControllers, likely won't exist.  

Same for all new, commented out XRI3 tests below.

AMollis replied:
To clarify the "existence" statement.  The controller class still exist, but will be deprecated.

I agree, rewriting this test will be required for a future PR.  Once the new rig is finished.

ms-RistoRK replied:
Ok, removed in https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/716/commits/85c93e1e40f889128b6ef6324ac17bd23a68dd39 , resolving.

whebertML commented on org.mixedrealitytoolkit.input/Tests/Runtime/BasicInputTests.cs:
```diff
@@ -592,6 +693,46 @@ public IEnumerator UntrackedControllerNearInteractions()
+            PropertyInfo[] accessorsInfos;
+            Type HandModel = typeof(HandModel);
+
+            fieldInfos = HandModel.GetFields(BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Public);
+            accessorsInfos = HandModel.GetProperties(BindingFlags.Instance | BindingFlags.Public);
+
+            var modelPrefabFieldInfo = fieldInfos.Where(fieldInfo => fieldInfo.Name.Equals("modelPrefab")).ToArray();
+            Assert.AreEqual(1, modelPrefabFieldInfo.Length, "HandModel is missing the 'modelPrefab' field");
+
+            var modelParentFieldInfo = fieldInfos.Where(fieldInfo => fieldInfo.Name.Equals("modelParent")).ToArray();
+            Assert.AreEqual(1, modelParentFieldInfo.Length, "HandModel is missing the 'modelParent' field");
+
+            var modelFieldInfo = fieldInfos.Where(fieldInfo => fieldInfo.Name.Equals("model")).ToArray();
+            Assert.AreEqual(1, modelFieldInfo.Length, "HandModel is missing the 'model' field");
+
+            var handNodeFieldInfo = fieldInfos.Where(fieldInfo => fieldInfo.Name.Equals("handNode")).ToArray();
+            Assert.AreEqual(1, modelFieldInfo.Length, "HandModel is missing the 'handNode' field");
+
+            var modelPrefabAccessorInfo = accessorsInfos.Where(accessorInfo => accessorInfo.Name.Equals("ModelPrefab")).ToArray();
+            Assert.AreEqual(1, modelPrefabAccessorInfo.Length, "HandModel is missing the 'ModelPrefab' accessor");
+
+            var modelParentAccessorInfo = accessorsInfos.Where(accessorInfo => accessorInfo.Name.Equals("ModelParent")).ToArray();
+            Assert.AreEqual(1, modelParentAccessorInfo.Length, "HandModel is missing the 'ModelParent' accessor");
+
+            var modelAccessorInfo = accessorsInfos.Where(accessorInfo => accessorInfo.Name.Equals("Model")).ToArray();
+            Assert.AreEqual(1, modelAccessorInfo.Length, "HandModel is missing the 'Model' accessor");
+
+            var handNodeAccessorInfo = accessorsInfos.Where(accessorInfo => accessorInfo.Name.Equals("HandNode")).ToArray();
+            Assert.AreEqual(1, handNodeAccessorInfo.Length, "HandModel is missing the 'HandNode' accessor");
+
```
What purpose does this test serve?  It is only testing the field and property layout of this component, not functionality.  This doesn't appear to be a standard type of test we typically write.

AMollis replied:
I see no harm in this.   I guess it could catch mistakenly deleted components.

ms-RistoRK replied:
I thought on this test aiming to ensure the deprecated XRController Model functionality is available in the XRI3 migrated code.  Also, I recall one bug in which there was a regression because of class having a field removed so this could help prevent that.  It is a "HandModel does have what it should" kind of test.  Resolving.

ms-RistoRK commented:
> > I am a doing research on VR frameworks and I really liked the MRTK approach in that it builds upon XRIT rather than trying to replace it. I'm impressed with what's been done in version 3.0 of XRIT and I'd love to see this PR come out as soon as possible. Do you think it's a matter of days/weeks/months? ;) @ms-RistoRK
> 
> Hello @IceTrooper, thank you for reaching out. We have it planned to release in 3 to 4 weeks, could be earlier though (I'm trying to make it happen asap). We are currently gathering internal feedback and doing extensive testing but everything is moving in the right direction yet we want to make sure before incorporating it into main branch. I'll ping you once it is in main branch.

@IceTrooper : I hope this finds you well.  Following up on letting you know when the XRI3 migration gets completed: I want to let you know that the last major PR has been approved and merged.  The controllerless XRI3+ compliant MRTK3 is available in the [feature/XRI3 branch](https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/commits/feature/XRI3/).  Please let us know if you find any issues :-)
